include/emitc/emitc_types.h

```diff
@@ -396,4 +396,46 @@ struct concat<Dim, T, Tensor4D<T, D0, D1, D2, D3>...> {
                               detail::first<sizeof...(D2)>({D2...}),
                               detail::sum<sizeof...(D3)>({D3...})>>>::type;
 };
+
+template <size_t... Shape>
+static constexpr std::array<size_t, sizeof...(Shape)> strides() {
+  std::array<size_t, sizeof...(Shape)> result = {};
+  constexpr std::array<size_t, sizeof...(Shape)> s = {Shape...};
+
+  if (sizeof...(Shape) == 0) {
+    return result;
+  }
+
+  result[sizeof...(Shape) - 1] = 1;
+
+  for (size_t i = sizeof...(Shape) - 1; i > 0; i--) {
+    result[i - 1] = result[i] * s[i];
+  }
+
+  return result;
+}
+
+template <size_t... Shape, typename... Indices>
+constexpr size_t ravel_index(Indices... indices) {
+  static_assert(sizeof...(Indices) == sizeof...(Shape),
+                "Incorrect number of arguments");
+
+  std::array<size_t, sizeof...(Shape)> shape = {Shape...};
+  std::array<size_t, sizeof...(indices)> indicesArray = {
+      static_cast<size_t>(indices)...};
+
+  for (size_t i = 0; i < sizeof...(Shape); ++i) {
+    assert(indicesArray[i] < shape[i]);
+  }
+
+  std::array<size_t, sizeof...(Shape)> s = strides<Shape...>();
+
+  size_t result = 0;
+  for (size_t i = 0; i < indicesArray.size(); ++i) {
+    result += indicesArray[i] * s[i];
+  }
+
+  return result;
+}
```

Comment:
Are these copied 1:1 from the Tensor class? If so we can remove the implementation from the class. Also  add `unravel_index` for completeness.

Reply:
Mostly, the parameters of the functions are different (`std::array` vs variadic arguments). That's why I didn't remove it from the class. Can we forward `std::array` to variadic?

Reply:
I'm pretty sure there is some meta programming voodoo you can do, or you could add an overload, or you keep it as is and we come back to this later.